### ISSUE TITLE
Error Linking CUDA during Install

### ISSUE THREAD

**Issue by RishabGargeya** [dmlc/mxnet#1385]
Hi,

I'm having an error linking -lcuda as I'm installing mxnet.
When I call make -j4 after modifying my config.mk file, I get the following:

```
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -shared -o lib/libmxnet.so build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/modules/Core/cuda/7.5/ -lcuda
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/modules/Core/cuda/7.5/ -lcuda
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
make: *** [lib/libmxnet.so] Error 1
make: *** Waiting for unfinished jobs....
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
make: *** [bin/im2rec] Error 1
```
Here is my config.mk file as well:
```
#-------------------------------------------------------------------------------
#  Template configuration for compiling mxnet
#
#  If you want to change the configuration, please use the following
#  steps. Assume you are on the root directory of mxnet. First copy the this
#  file so that any local changes will be ignored by git
#
#  $ cp make/config.mk .
#
#  Next modify the according entries, and then compile by
#
#  $ make
#
#  or build in parallel with 8 threads
#
#  $ make -j8
#-------------------------------------------------------------------------------

#---------------------
# choice of compiler
#--------------------

export CC = gcc
export CXX = g++
export NVCC = nvcc

# whether compile with debug
DEBUG = 1

# the additional link flags you want to add
ADD_LDFLAGS = -L/share/sw/modules/Core/cuda/7.5/

# the additional compile flags you want to add
ADD_CFLAGS =

#---------------------------------------------
# matrix computation libraries for CPU/GPU
#---------------------------------------------

# whether use CUDA during compile
USE_CUDA = 1

# add the path to CUDA library to link and compile flag
# if you have already add them to environment variable, leave it as NONE
# USE_CUDA_PATH = /usr/local/cuda
USE_CUDA_PATH = NONE

# whether use CUDNN R3 library
USE_CUDNN = 1

# whether use cuda runtime compiling for writing kernels in native language (i.e. Python)
USE_NVRTC = 0

# whether use opencv during compilation
# you can disable it, however, you will not able to use
# imbin iterator
USE_OPENCV = 1

# use openmp for parallelization
USE_OPENMP = 0

# choose the version of blas you want to use
# can be: mkl, blas, atlas, openblas
USE_BLAS = openblas

# add path to intel library, you may need it for MKL, if you did not add the path
# to environment variable
USE_INTEL_PATH = NONE

#----------------------------
# distributed computing
#----------------------------

# whether or not to enable multi-machine supporting
USE_DIST_KVSTORE = 0

# whether or not allow to read and write HDFS directly. If yes, then hadoop is
# required
USE_HDFS = 0

# path to libjvm.so. required if USE_HDFS=1
LIBJVM=$(JAVA_HOME)/jre/lib/amd64/server

# whether or not allow to read and write AWS S3 directly. If yes, then
# libcurl4-openssl-dev is required, it can be installed on Ubuntu by
# sudo apt-get install -y libcurl4-openssl-dev
USE_S3 = 0

#----------------------------
# additional operators
#----------------------------

# path to folders containing projects specific operators that you don't want to put in src/operators
EXTRA_OPERATORS =


#----------------------------
# plugins
#----------------------------

# whether to use torch integration. This requires installing torch.
# TORCH_PATH = $(HOME)/torch
# MXNET_PLUGINS += plugin/torch/torch.mk
```
Any thoughts? Cuda is installed in /share/sw/modules/Core/cuda/7.5/.
Thanks for the help.

**Comment by piiswrong** [comment]
You have USE_CUDA_PATH = NONE in your config.mk. Change it to USE_CUDA_PATH = /share/sw/modules/Core/cuda/7.5/

**Comment by RishabGargeya** [comment]
Thanks for the quick response. I tried that and here is the error:

```
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/share/sw/modules/Core/cuda/7.5//include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -shared -o lib/libmxnet.so build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -L/share/sw/modules/Core/cuda/7.5//lib64 -L/share/sw/modules/Core/cuda/7.5//lib -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/modules/Core/cuda/7.5/ -lcuda
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/share/sw/modules/Core/cuda/7.5//include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -L/share/sw/modules/Core/cuda/7.5//lib64 -L/share/sw/modules/Core/cuda/7.5//lib -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/modules/Core/cuda/7.5/ -lcuda
cc1plus: error: /share/sw/modules/Core/cuda/7.5//include: Not a directory
make: *** [bin/im2rec] Error 1
make: *** Waiting for unfinished jobs....
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
make: *** [lib/libmxnet.so] Error 1
```
All I did was change USE_CUDA_PATH = NONE to USE_CUDA_PATH = /share/sw/modules/Core/cuda/7.5/

**Comment by piiswrong** [comment]
do you have the directory /share/sw/modules/Core/cuda/7.5/include? Try /share/sw/modules/Core/cuda/7.5/ -> /share/sw/modules/Core/cuda/7.5

**Comment by RishabGargeya** [comment]
Hi,
I checked in /share/sw/modules/Core/cuda/7.5/ and apparently I have the wrong directory as it does not have an include folder.
CUDA on my system is actually installed to "/share/sw/free/cuda/7.5".
I added that folder to my config.mk to be the following;

```
#-------------------------------------------------------------------------------
#  Template configuration for compiling mxnet
#
#  If you want to change the configuration, please use the following
#  steps. Assume you are on the root directory of mxnet. First copy the this
#  file so that any local changes will be ignored by git
#
#  $ cp make/config.mk .
#
#  Next modify the according entries, and then compile by
#
#  $ make
#
#  or build in parallel with 8 threads
#
#  $ make -j8
#-------------------------------------------------------------------------------

#---------------------
# choice of compiler
#--------------------

export CC = gcc
export CXX = g++
export NVCC = nvcc

# whether compile with debug
DEBUG = 1

# the additional link flags you want to add
ADD_LDFLAGS = -L/share/sw/free/cuda/7.5

# the additional compile flags you want to add
ADD_CFLAGS =

#---------------------------------------------
# matrix computation libraries for CPU/GPU
#---------------------------------------------

# whether use CUDA during compile
USE_CUDA = 1

# add the path to CUDA library to link and compile flag
# if you have already add them to environment variable, leave it as NONE
# USE_CUDA_PATH = /usr/local/cuda
USE_CUDA_PATH = /share/sw/free/cuda/7.5

# whether use CUDNN R3 library
USE_CUDNN = 1

# whether use cuda runtime compiling for writing kernels in native language (i.e. Python)
USE_NVRTC = 0

# whether use opencv during compilation
# you can disable it, however, you will not able to use
# imbin iterator
USE_OPENCV = 1

# use openmp for parallelization
USE_OPENMP = 0

# choose the version of blas you want to use
# can be: mkl, blas, atlas, openblas
USE_BLAS = openblas

# add path to intel library, you may need it for MKL, if you did not add the path
# to environment variable
USE_INTEL_PATH = NONE

#----------------------------
# distributed computing
#----------------------------

# whether or not to enable multi-machine supporting
USE_DIST_KVSTORE = 0

# whether or not allow to read and write HDFS directly. If yes, then hadoop is
# required
USE_HDFS = 0

# path to libjvm.so. required if USE_HDFS=1
LIBJVM=$(JAVA_HOME)/jre/lib/amd64/server

# whether or not allow to read and write AWS S3 directly. If yes, then
# libcurl4-openssl-dev is required, it can be installed on Ubuntu by
# sudo apt-get install -y libcurl4-openssl-dev
USE_S3 = 0

#----------------------------
# additional operators
#----------------------------

# path to folders containing projects specific operators that you don't want to put in src/operators
EXTRA_OPERATORS =


#----------------------------
# plugins
#----------------------------

# whether to use torch integration. This requires installing torch.
# TORCH_PATH = $(HOME)/torch
# MXNET_PLUGINS += plugin/torch/torch.mk
```

I also modified my LD_LIBRARY_PATH, PATH, and LIBRARY_PATH:
```
export CUDA_HOME=/share/sw/free/cuda/7.5
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${CUDA_HOME}/lib64
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${CUDA_HOME}/lib
export LIBRARY_PATH=$LIBRARY_PATH:${CUDA_HOME}/lib64
export LIBRARY_PATH=$LIBRARY_PATH:${CUDA_HOME}/lib
export PATH=$PATH:/share/sw/free/cuda/7.5/
```

This is the error I am still receiving:
```
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/share/sw/free/cuda/7.5/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -shared -o lib/libmxnet.so build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -L/share/sw/free/cuda/7.5/lib64 -L/share/sw/free/cuda/7.5/lib -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/free/cuda/7.5 -lcuda
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/share/sw/free/cuda/7.5/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -L/share/sw/free/cuda/7.5/lib64 -L/share/sw/free/cuda/7.5/lib -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/free/cuda/7.5 -lcuda
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
make: *** [lib/libmxnet.so] Error 1
make: *** Waiting for unfinished jobs....
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
make: *** [bin/im2rec] Error 1
```

**Comment by piiswrong** [comment]
When linker cannot find a lib, find that lib and add its path to linker
path.

**Comment by RishabGargeya** [comment]
Hi,
Thanks for the reply.
How would I go about doing this? Adding paths to the LDFLAGS environment variable, LD_LIBRARY_PATH and LIBRARY_PATH didn't seem to help before.

**Comment by RishabGargeya** [comment]
Hi,
I tried to change the LDFLAGS variable to the directory immediately containing libcuda.so 
(-L/share/sw/free/cuda/7.5/lib64/stubs/).
I believe it is now recognizing the library, but I encounter another error:

```
g++ -DMSHADOW_FORCE_STREAM -Wall -g -O0 -DDMLC_LOG_FATAL_THROW=0 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/share/sw/free/cuda/7.5/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 `pkg-config --cflags opencv` -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/common/tblob_op_registry.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/ndarray/unary_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/leaky_relu.o build/src/operator/lrn.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o dmlc-core/libdmlc.a build/src/ndarray/ndarray_function_gpu.o build/src/ndarray/unary_function_gpu.o build/src/operator/activation_gpu.o build/src/operator/batch_norm_gpu.o build/src/operator/block_grad_gpu.o build/src/operator/cast_gpu.o build/src/operator/concat_gpu.o build/src/operator/convolution_gpu.o build/src/operator/crop_gpu.o build/src/operator/cudnn_batch_norm_gpu.o build/src/operator/deconvolution_gpu.o build/src/operator/dropout_gpu.o build/src/operator/elementwise_binary_op_gpu.o build/src/operator/elementwise_binary_scalar_op_gpu.o build/src/operator/elementwise_sum_gpu.o build/src/operator/embedding_gpu.o build/src/operator/fully_connected_gpu.o build/src/operator/identity_attach_KL_sparse_reg_gpu.o build/src/operator/leaky_relu_gpu.o build/src/operator/lrn_gpu.o build/src/operator/native_op_gpu.o build/src/operator/pooling_gpu.o build/src/operator/regression_output_gpu.o build/src/operator/reshape_gpu.o build/src/operator/slice_channel_gpu.o build/src/operator/softmax_activation_gpu.o build/src/operator/softmax_output_gpu.o build/src/operator/swapaxis_gpu.o build/src/operator/upsampling_gpu.o build/src/optimizer/sgd_gpu.o -pthread -lm -lcudart -lcublas -lcurand -L/share/sw/free/cuda/7.5/lib64 -L/share/sw/free/cuda/7.5/lib -lopenblas -lrt `pkg-config --libs opencv` -lcudnn -L/share/sw/free/cuda/7.5/lib64/stubs/ -lcuda
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: dmlc-core/libdmlc.a(data.o): undefined reference to symbol 'omp_get_num_threads@@OMP_1.0'
//usr/lib64/libgomp.so.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make: *** [bin/im2rec] Error 1
```
Does anything I am doing look wrong?
What do you think omp_get_num_threads@@OMP_1.0 and DSO missing from command line mean?

**Comment by piiswrong** [comment]
Try turn off OpenMP support in config.mk.

**Comment by RishabGargeya** [comment]
Hi,
Thanks for the reply. I actually tried turning on OpenMP support and mxnet successfully compiled.
I continued to successfully install python support through ```python setup.py develop --user```.
However, I am unable to import mxnet - It still gives me issues with recognizing the cuda library:
Here is the error when I execute import mxnet:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/rgargeya/mxnet/python/mxnet/__init__.py", line 7, in <module>
    from .base import MXNetError
  File "/home/rgargeya/mxnet/python/mxnet/base.py", line 43, in <module>
    _LIB = _load_lib()
  File "/home/rgargeya/mxnet/python/mxnet/base.py", line 35, in _load_lib
    lib = ctypes.cdll.LoadLibrary(lib_path[0])
  File "/home/rgargeya/miniconda2/envs/highway/lib/python2.7/ctypes/__init__.py", line 443, in LoadLibrary
    return self._dlltype(name)
  File "/home/rgargeya/miniconda2/envs/highway/lib/python2.7/ctypes/__init__.py", line 365, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcuda.so.1: cannot open shared object file: No such file or directory
```
MXNET is still unable to find libcuda.so but I added the path of the folder that directly contains the file to LD_LIBRARY_PATH, and mxnet compiled successfully beforehand? Why is it not able to find the file during runtime? Thanks for all your help.

**Comment by shenxiul** [comment]
Hi @RishabGargeya , I got the same issue and fixed it. Are you compiling on a machine that actually have a GPU? Or you're just compiling with CUDA library (which is usually the case in super computers)? It was not looking for libcuda.so but libcuda.so.1, which should locate in /usr/local/lib. (Try `ldd lib/libmxnet.so and you will see it`) You have to compile it and run it on a machine with a GPU on it.

**Comment by RishabGargeya** [comment]
Hi @shenxiul, I am indeed running mxnet using shared resources on a supercomputer. I was actually able to run mxnet when directly accessing a gpu node but was unable to call ```import mxnet``` when on a cpu (getting the above error). I compiled mxnet as above and was hoping to conduct some tests on a cpu independently. How were you able to fix the issue? The /usr/local/lib directory seems to be empty and I don't have permission to write to it.

**Comment by shenxiul** [comment]
@RishabGargeya , this is due to the implementation of mxnet. Once we compile it with CUDA option, whenever we are loading libmxnet.so, it has to find libcuda.so.1, in other words it must be run on a node with GPU on it. We can still use CPU mode to let it runs on CPU only, but must on a machine that has GPU installed. That's might be a potential flaw, can we open another issue or PR stating this problem and close this issue?

**Comment by RishabGargeya** [comment]
Hi @shenxiul, of course. We can close this issue and open a PR.
